src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -777,7 +773,9 @@ public function expects_closer( $node = null ): ?bool {
 			// Void elements.
 			self::is_void( $token_name ) ||
 			// Special atomic elements.
-			in_array( $token_name, array( 'IFRAME', 'NOEMBED', 'NOFRAMES', 'SCRIPT', 'STYLE', 'TEXTAREA', 'TITLE', 'XMP' ), true )
+			in_array( $token_name, array( 'IFRAME', 'NOEMBED', 'NOFRAMES', 'SCRIPT', 'STYLE', 'TEXTAREA', 'TITLE', 'XMP' ), true ) ||
+			// Self-closing elements in foreign content.
+			( isset( $node ) && 'html' !== $node->namespace && $node->has_self_closing_flag )
```

Comment:
Unless there is a performance benefit for doing this, I think generally better to check for equality with `null` since the variable _is_ set. But even better here to check if it is the instance of the expected class, right?

```suggestion
			( $node instanceof WP_HTML_Token && 'html' !== $node->namespace && $node->has_self_closing_flag )
```

Reply:
yeah. I think I did this for a reason, but I can't remember. likely I had other code in transition and had non-null non-token instances passed around

Reply:
Bookmarking this, the `isset( $node )` seems like it would make this function return `true` under some circumstances where we would not expect that.

Reply:
there was much discussion over this little function, and a lot of effort to "fix" it that only resulted in broken code.

I'd like to propose we continue with it as written and come back afterwards to clean it up.

right now, all of the "fixes" are only used in test code when `expects_closer()` is called without a `$node` passed in. the real problem is that there's confusion in what this function is communicating and how `step()` pops elements off of the stack of open elements.

we recently realized that we need to pop elements off not in `step()`, but in the parsing rules where they indicate as much. I'd rather we ship this odd bit now and create the more-sweeping fix in one go instead of stewing on code we think must be wrong but which we don't have a failing test case for.

---

I've added a type annotation to the function, ensuring that `isset()` is sufficient, and I've added a `@todo` annotation earmarking this for review. I _think_ what's going on is a complicated interaction with `step()` where this looks wrong but isn't.

By the way, this function returns "expects closer" for all closing tags, which is "wrong" but also part of the bigger issue not created in this PR.

Thoughts?

Reply:
PR to address this: https://github.com/WordPress/wordpress-develop/pull/7162